### ISSUE TITLE
Add related error spans for getter/setters with different types

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#25002]
Now that we support related spans (#10489), we'd like to improve an existing error message.

Currently, we provide a diagnostic for a pair of `get`/`set` accessor's types not matching:

Code:

```ts
let x = {
  get foo() { return 100; }
  set foo(value: string): { }
}
```

Current error:

```
'get' and 'set' accessor must have the same type.
```

We'd like to give a better error message. For example:

Primary span:

```
A 'get-' and 'set-' accessor must have the same type, but this 'get' accessor has the type '{0}'.
```

Related span:

```
The respective 'set' accessor has the type '{0}'.
```

**Comment by bashdx** [comment]
Hello guys, 

if you could point me in the correct direction (where do I find the error messages) I will take a look at it.
Thank you and regards
bashdx

**Comment by mhegazy** [comment]
The issue in the OP is about adding additional information to the error message. We have recently allowed an error message to include child messages (to add clarifications), we call these related spans. See https://github.com/Microsoft/TypeScript/pull/25359 for a similar change.

**Comment by bashdx** [comment]
Thank you for your reply. I built the compiler and tried the following with your code snippet from above:
`$ node tsc.js --target "ES5" test_accessors.ts 
test_accessors.ts:4:3 - error TS2322: Type '100' is not assignable to type 'string'.

4  	return 100;
   	~~~~~~~~~~~

test_accessors.ts:6:26 - error TS1005: '{' expected.

6  set foo(value: string):{}
                           ~
`
I can't get error `2380` to be invoked. Could you please provide a fitting code snippet. Using a class instead of an object resulted in a similar result.

Thank you and regards

**Comment by mhegazy** [comment]
Try this instead:

```ts
let x = {
    get foo(): number { return 100; },
    set foo(value: string) { }
}
```

**Comment by bashdx** [comment]
That did it! I am getting closer - could you please tell me, how to create a "blank" node? The current result looks like this (please ignore the unrelated error message about generators - this was solely for testing). Where would I create the message for the related span?

```
built/local/test_accessors.ts:2:9 - error TS2380: 'get' and 'set' accessor must have the same type, but this 'get' accessor has the type 'number'.

2     get foo(): number { return 100; },
          ~~~

  built/local/test_accessors.ts:3:9
    3     set foo(value: string) { }
              ~~~
    Generators are not allowed in an ambient context.

built/local/test_accessors.ts:3:9 - error TS2380: 'get' and 'set' accessor must have the same type, but this 'get' accessor has the type 'string'.

3     set foo(value: string) { }
          ~~~

  built/local/test_accessors.ts:2:9
    2     get foo(): number { return 100; },
              ~~~
    Generators are not allowed in an ambient context.

```

**Comment by mhegazy** [comment]
not sure why you are getting the error `Generators are not allowed in an ambient context.`..

**Comment by bashdx** [comment]
The rest of the output can stay like this? This very format? Also it seems the error handler is invoked twice, while the former aims for the getter (inteded) and the latter aims for the setter, but uses the getter error message. I will check it out ASAP.

**Comment by bashdx** [comment]
I'd appreciate feedback if everything went fine, please.
Find pull request [here](https://github.com/Microsoft/TypeScript/pull/25415).

**Comment by RyanCavanaugh** [comment]
Open to ideas here that have some concrete improvement

**Comment by acagastya** [comment]
There is no restriction on `get` `set` in ECMA specification, so adhering to the standard, shouldn't this restriction be lifted?  I mean "'get' and 'set' accessor must have the same type".

**Comment by sirian** [comment]
![image](https://user-images.githubusercontent.com/897643/77229064-8a71b780-6b9c-11ea-95a8-0c26472a8f7c.png)

setters and getters MAY HAVE different types

**Comment by sirian** [comment]
```javascript
get zIndex(): string
set zIndex(value: string | number | null | undefined)
```

**Comment by martinheidegger** [comment]
To me its a reasonable limitation that getter/setter should not have different type but this should imo. be handled by a linter rather than the compiler.

What I have been thinking is that this might be related to [the writeonly proposal #21759](https://github.com/microsoft/TypeScript/issues/21759). Maybe having a type declaration such as the following could help to get a mental model for how getters/setters have different types:

```typescript
interface IMY_URL {
  readonly query: Record<string, string>
  writeonly query: string | Record <string, string | number | null | undefined>
}
```

Another real-life example I have for this is the [sketch api](https://developer.sketch.com/reference/api/#style) which does C mappings and has properties that can be set with partial objects but the getters will return filled-out instances.

**Comment by acagastya** [comment]
@martinheidegger that limitation is not per ECMA standards and only serves as additional gatekeeping.

**Comment by szmarczak** [comment]
I completely agree with @acagastya and @sirian. Different setters/getters are very useful when doing normalization.

/cc @sindresorhus
